sheep/src/pack/maxrects.rs

```diff
@@ -0,0 +1,369 @@
+use {Packer, PackerResult, SpriteAnchor, SpriteData};
+
+pub struct MaxrectsPacker;
+
+#[derive(Copy, Clone)]
+pub struct MaxrectsOptions {
+    preferred_width: u32,
+    preferred_height: u32,
+}
+
+impl Default for MaxrectsOptions {
+    fn default() -> Self {
+        MaxrectsOptions {
+            preferred_width: 4096,
+            preferred_height: 4096,
+        }
+    }
+}
+
+impl MaxrectsOptions {
+    pub fn preferred_width(mut self, width: u32) -> Self {
+        self.preferred_width = width;
+        self
+    }
+
+    pub fn preferred_height(mut self, height: u32) -> Self {
+        self.preferred_height = height;
+        self
+    }
+}
+
+impl Packer for MaxrectsPacker {
+    type Options = MaxrectsOptions;
+
+    fn pack(sprites: &[SpriteData], options: MaxrectsOptions) -> Vec<PackerResult> {
+        let mut bins = Vec::new();
+        let mut oversized = Vec::new();
+
+        // First, filter out all oversized sprites
+        let mut sprites = sprites
+            .iter()
+            .enumerate()
+            .filter(|(i, sprite)| {
+                if sprite.dimensions.0 > options.preferred_width
+                    || sprite.dimensions.1 > options.preferred_height
+                {
+                    oversized.push(MaxRectsBin::oversized(sprite.dimensions, *i));
+                    false
+                } else {
+                    true
+                }
+            })
+            .map(|(_, sprite)| *sprite)
+            .collect::<Vec<_>>();
+
+        // Now, keep inserting as many as possible into each bin until
+        // all sprites have been placed. Since all oversized rects have
+        // already been filtered out, this will always terminate.
+        while !sprites.is_empty() {
+            let mut bin = MaxRectsBin::new(options.preferred_width, options.preferred_height);
+            sprites = bin.insert_sprites(&sprites);
+            bins.push(bin);
+        }
+
+        bins.extend(oversized.into_iter());
+        let result = bins.into_iter().map(|bin| bin.to_result()).collect::<Vec<PackerResult>>();
+
+        result
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+struct Rect {
+    pub x: u32,
+    pub y: u32,
+    pub width: u32,
```

Comment:
looks like most of the code would greatly benefit from different rect representation, allowing greater performance and less complexity

```rs
struct Rect {
    pub min_x: u32,
    pub min_y: u32,
    pub max_x: u32,
    pub max_y: u32,
}
```

`contains` check can then be made stupidly fast :P